core/src/net/buffers/chunk_ref.rs

```diff
@@ -0,0 +1,233 @@
+use crate::prelude::Buf;
+use std::sync::Arc;
+
+/// A ChunkRef is a Byte-buffer which locks the BufferChunk which the ChunkRef is a lease from.
+/// It's a cloneable read-only equivalent of ChunkLease.
+#[derive(Debug, Clone)]
+pub struct ChunkRef {
+    content: &'static [u8],
+    read_pointer: usize,
+    chain_head_len: usize,
+    lock: Arc<u8>,
+    chain: Option<Box<ChunkRef>>,
+    /// The length of the chain from self (i.e. independent of parent(s))
+    chain_len: usize,
+}
+
+impl ChunkRef {
+    pub(crate) fn new(
+        content: &'static [u8],
+        read_pointer: usize,
+        chain_head_len: usize,
+        lock: Arc<u8>,
+        chain: Option<Box<ChunkRef>>,
+        chain_len: usize,
+    ) -> Self {
+        Self {
+            content,
+            read_pointer,
+            chain_head_len,
+            lock,
+            chain,
+            chain_len,
+        }
+    }
+
+    /// The full length of the underlying bytes without read/write-pointers
+    /// Use `remaining()`/`remaining_mut()` for readable/writable lengths
+    pub fn capacity(&self) -> usize {
+        self.chain_len
+    }
+
+    /// Appends `new_tail` to the end of the ChunkRef chain
+    /// Use Public ChunkLease methods into_chunk_ref_as_tail or into_chunk_ref_with_tail
+    pub(super) fn chain(&mut self, new_tail: ChunkRef) {
+        self.chain_len += new_tail.chain_len;
+        if let Some(tail) = &mut self.chain {
+            // recursion
+            tail.chain(new_tail)
+        } else {
+            // recursion complete
+            self.chain = Some(Box::new(new_tail))
+        }
+    }
+
+    // Recursive method for the bytes() impl
+    fn get_bytes_at(&self, pos: usize) -> &[u8] {
+        if pos >= self.chain_head_len {
+            if let Some(chain) = &self.chain {
+                chain.get_bytes_at(pos - self.chain_head_len)
+            } else {
+                panic!("Critical Bug in ChunkRef, bad chain");
+            }
+        } else {
+            let slice: &[u8] = &*self.content;
+            &slice[pos..]
+        }
+    }
+}
+
+impl Buf for ChunkRef {
+    fn remaining(&self) -> usize {
+        self.chain_len - self.read_pointer
+    }
+
+    // Returns a slice starting from the read-pointer up to (possibly) the
+    fn bytes(&self) -> &[u8] {
+        self.get_bytes_at(self.read_pointer)
+    }
+
+    fn advance(&mut self, cnt: usize) {
+        self.read_pointer += cnt;
+    }
+}
+
+unsafe impl Send for ChunkRef {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::net::buffers::{BufferConfig, EncodeBuffer};
+    use bytes::{BufMut, Bytes};
+
+    /// Use different data sizes and buffer sizes to test different kinds of chains/splits.
+    /// We want to test Chaining, Creation functions from chunk_lease.rs and reading/using Buf impl
+    /// Variants:
```

Comment:
I don't think it makes sense to use doc comments for test functions. Should probably just use normal comments